solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpSolrJdkClient.java

```diff
@@ -0,0 +1,453 @@
+import java.net.InetSocketAddress;
+import java.net.ProxySelector;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.net.http.HttpTimeoutException;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.Collection;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import javax.net.ssl.SSLContext;
+import org.apache.solr.client.solrj.ResponseParser;
+import org.apache.solr.client.solrj.SolrRequest;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.request.RequestWriter;
+import org.apache.solr.client.solrj.util.ClientUtils;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.apache.solr.common.util.ContentStream;
+import org.apache.solr.common.util.ExecutorUtil;
+import org.apache.solr.common.util.NamedList;
+import org.apache.solr.common.util.ObjectReleaseTracker;
+import org.apache.solr.common.util.SolrNamedThreadFactory;
+import org.eclipse.jetty.util.BlockingArrayQueue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class HttpSolrJdkClient extends HttpSolrClientBase {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private static final String USER_AGENT =
+      "Solr[" + MethodHandles.lookup().lookupClass().getName() + "] 1.0";
+
+  protected HttpClient client;
+
+  protected ExecutorService executor;
+
+  private boolean shutdownExecutor;
+
+  protected HttpSolrJdkClient(String serverBaseUrl, HttpSolrJdkClient.Builder builder) {
+    super(serverBaseUrl, builder);
+
+    HttpClient.Redirect followRedirects =
+        Boolean.TRUE.equals(builder.followRedirects)
+            ? HttpClient.Redirect.NORMAL
+            : HttpClient.Redirect.NEVER;
+    HttpClient.Builder b = HttpClient.newBuilder().followRedirects(followRedirects);
+    if (builder.sslContext != null) {
+      b.sslContext(builder.sslContext);
+    }
+
+    if (builder.executor != null) {
+      this.executor = builder.executor;
+      this.shutdownExecutor = false;
+    } else {
+      BlockingArrayQueue<Runnable> queue = new BlockingArrayQueue<>(32, 32);
+      this.executor =
+          new ExecutorUtil.MDCAwareThreadPoolExecutor(
+              4,
+              256,
+              60,
+              TimeUnit.SECONDS,
+              queue,
+              new SolrNamedThreadFactory(this.getClass().getSimpleName()));
+      this.shutdownExecutor = true;
+    }
+    b.executor(this.executor);
+
+    if (builder.useHttp1_1) {
+      b.version(HttpClient.Version.HTTP_1_1);
```

Comment:
If you expressly set the version to HTTP 2, can we avoid the ping altogether?

Reply:
I would rather we avoid the ping at the expense of the possibility that a minority of users would be required to configure HTTP 1.1 explicitly on their client.  After all, we know Solr supports Http/2 and it's been out there for quite some time.  I can understand why the JDK's HttpClient defaults the way it does (kind of in the middle) but we know how we deploy Solr for the majority of users.

Reply:
We can't tell it to always use Http/2.  Although this feature was requested (see https://bugs.openjdk.org/browse/JDK-8287589) it does not look like the JDK client will support this any time soon.